pgx/src/datum/range.rs

```diff
@@ -0,0 +1,371 @@
+        }
+    }
+
+    /// Builds an "empty" range
+    pub fn empty_range_data() -> Self {
+        let lower_bound = RangeBound { lower: true, ..RangeBound::default() };
+        let upper_bound = RangeBound { lower: false, ..RangeBound::default() };
+        Self::from_range_bounds_internal(lower_bound, upper_bound, true)
+    }
+
+    /// Generate a RangeData<T> from the lower/upper RangeBounds, implies non-empty
+    #[inline]
+    pub fn from_range_bounds(lower_bound: RangeBound, upper_bound: RangeBound) -> Self {
+        Self::from_range_bounds_internal(lower_bound, upper_bound, false)
+    }
+
+    pub(crate) fn from_range_bounds_internal(
+        lower_bound: RangeBound,
+        upper_bound: RangeBound,
+        is_empty: bool,
+    ) -> Self {
+        RangeData { lower: lower_bound, upper: upper_bound, is_empty, __marker: PhantomData }
+    }
+
+    /// Generate a RangeData<T> from the T values for lower/upper bounds, lower/upper inclusive
+    /// None for lower_val or upper_val will represent lower_inf/upper_inf bounds
+    pub fn from_range_values(
+        lower_val: Option<T>,
+        upper_val: Option<T>,
+        lower_inc: bool,
+        upper_inc: bool,
+    ) -> Self {
+        let mut lower_bound =
+            RangeBound { lower: true, inclusive: lower_inc, ..Default::default() };
+
+        let mut upper_bound =
+            RangeBound { lower: false, inclusive: upper_inc, ..Default::default() };
+
+        match lower_val {
+            Some(lower_val) => {
+                lower_bound.val =
+                    lower_val.into_datum().expect("Couldn't convert lower_val to Datum");
+            }
+            None => {
+                lower_bound.infinite = true;
+            }
+        }
+
+        match upper_val {
+            Some(upper_val) => {
+                upper_bound.val =
+                    upper_val.into_datum().expect("Couldn't convert upper_val to Datum");
+            }
+            None => {
+                upper_bound.infinite = true;
+            }
+        }
+
+        RangeData::from_range_bounds(lower_bound, upper_bound)
+    }
+}
+
+impl<T> From<Range<T>> for RangeData<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    /// ## Safety
+    /// Requires that:
+    /// - range.range_type is valid pointer to RangeType PG struct
+    fn from(range: Range<T>) -> Self {
+        let mut lower_bound: RangeBound = Default::default();
+        let mut upper_bound: RangeBound = Default::default();
+        let mut is_empty = false;
+
+        unsafe {
+            let typecache = pg_sys::lookup_type_cache(
+                (*(range.range_type)).rangetypid,
+                pg_sys::TYPECACHE_RANGE_INFO as i32,
+            );
+
+            pg_sys::range_deserialize(
+                typecache,
+                range.range_type,
+                &mut lower_bound,
+                &mut upper_bound,
+                &mut is_empty,
+            );
+        }
```

Comment:
unsafe block, preconditions, postconditions, etc., you can keep it brief but it'd be nice.